### PR TITLE
songLyrics extension

### PR DESCRIPTION
**Pull request by kgarner7**
Seeks to implement https://github.com/opensubsonic/open-subsonic-api/discussions/60
The representation is slightly different (`lyricsList` > `structuredLyrics`), but I believe that should be friendlier for XML (and also differentiate from the existing `lyrics` object)

### CODE REVIEW

**Comment by dweymouth** [comment]
Can you add example XML as well at least in the PR description? Going forward I think we should have example XML output for all the OpenSubsonic responses as well but that can be a separate task

File: content/en/docs/Responses/lyricsList.md
Comment (dweymouth) [inline content/en/docs/Responses/lyricsList.md:64]: synced="false"

File: content/en/docs/Responses/structuredLyrics.md
Comment (dweymouth) [inline content/en/docs/Responses/structuredLyrics.md:63]: synced="false"

**Comment by dweymouth** [comment]
I believe we want a doc in the Endpoints directory too, the title postfixed with [OS] to identify it as an OpenSubsonic endpoint

**Comment by Tolriq** [comment]
As @dweymouth said the extension page just define the extension names and what they enable with link to the actual endpoints / responses see https://deploy-preview-62--opensubsonic.netlify.app/docs/extensions/template/

This is done like that, because an extension can include 3 endoints for example and 3 changes to current end points this will be a mess to have all the in the extension directly.

`structured-lyrics` shoud be camelCase so `structuredLyrics`. Except the "subsonic-response" every other field respect that.

We should also expose an `offset` value at the root of `structuredLyrics` 

See offset https://en.wikipedia.org/wiki/LRC_(file_format) that exist in LRC format.

In `structuredLyrics` I do not think

```
          "artist": "Muse",
          "artistId": "1234",
```
makes sense anymore as it's not OpenSubsonic multiple artist / album artists style.

I would only return a `displayArtist` value and since we request by `songId` I would return it too so that client can ensure they get the correct one.

And lastly since we can expose multiple lyrics the `title` value is maybe not clear enough that it's the song title and not a title value for the lyrics specially when LRC files can have a title tag that may not match the actual song title as may be tied to the song language.

I'd say `songTitle` and `lyricsTitle` fields if we want to expose both.

File: content/en/docs/Responses/structuredLyrics.md
Comment (dweymouth) [inline content/en/docs/Responses/structuredLyrics.md:81]: Why do we need this field? Do lyrics sources sometimes return the timestamps misaligned? If so, could the server pre-process the lyrics and update the `start` timestamp for all the lines to take care of the offset?

Comment (Tolriq) [reply content/en/docs/Responses/structuredLyrics.md:81]: This is a field of LRC often used to perfectly sync to the user liking and it allows the client to know it's an aligned lyric and so may decide to not apply it's default internal offset in those cases.

**Review by dweymouth — APPROVED** [review]

**Comment by Tolriq** [comment]
@kgarner7 there's should not be different kind of OpenSubsonic example tabs since the responses must be always the same.
The differences must be based on parameters optional or not.
The different parameters optional or not and the documentation must be enough to clear any doubts.
But the example should always cover all fields to be able to build code from it for clients with strict answers handling.

Unfortunately due to the mess of Subsonic API we can't publish a proper schema that would help, so the doc is the schema.

**Comment by dweymouth** [comment]
@Tolriq I'm not sure what you mean by not having different OpenSubsonic tabs as it seems we've already done this in production: https://opensubsonic.netlify.app/docs/endpoints/getalbum/

I think the value is to show what things were added in the OpenSubsonic response compared to the legacy one, but you're right that it should be backward compatible, so if we want we can not have separate tabs for Subsonic and OpenSubsonic and just do separate tabs for JSON and XML. But if we want to go this way, we should clean up the docs already in production and get rid of the OpenSubsonic tabs

**Comment by Tolriq** [comment]
Check the page before ;)

He added 2 OpenSubsonic Json and 2 XML ones. There should be only one.

I wrote a long text to explain why.

**Comment by dweymouth** [comment]
> Check the page before ;)
> 
> He added 2 OpenSubsonic Json and 2 XML ones. There should be only one.
> 
> I wrote a long text to explain why.

my bad

**Comment by Tolriq** [comment]
@kgarner7 there's no need for multiple examples, the doc explains all that is needed. All the other pages don't show all the possible variations with the optional fields, this makes it hard to read and generate a schema in client apps.

File: content/en/docs/Responses/structuredLyrics.md
Comment (Tolriq) [inline content/en/docs/Responses/structuredLyrics.md:55]: synced should be removed.

**Comment by kgarner7** [comment]
> @kgarner7 there's no need for multiple examples, the doc explains all that is needed. All the other pages don't show all the possible variations with the optional fields, this makes it hard to read and generate a schema in client apps.

With all due respect, I think it is important at least for structured lyrics to show that it must either be **all** synced lines, or **all** unsynced. If you have a better way of phrasing that that doesn't show both variants, then please go ahead

**Comment by Tolriq** [comment]
@kgarner7 the example field is supposed to be a valid JSON, by putting 2 JSON in it this is no more valid JSON and it makes it hard to read and understand.

A schema is a schema and we do not support nor want to embrace polymorphism. So while it's not recommended the API and the schema does not prevent a synced lyrics to not return start (And you embrace that in the doc by saying it's assumed it's 0 ...)
And nothing prevent an unsynced lyrics to provide start, the client should just ignore the value as assumed it's wrong.

The documentation about the field should be enough. Double examples does not bring anything here and just makes things harder and the doc less clean.

Specially when you say it should be present for structured lyrics when the only change that actually occurs is in line part.

And your doc saying "The start time of the lyrics, in milliseconds. If this is not part of synced lyrics, start must be omitted" is quite clear.

So sorry but again no there's no need to duplicate the examples.

See https://deploy-preview-62--opensubsonic.netlify.app/docs/responses/contributor/ there's only 1 example covering all the fields. Not multiple ones to show when the subRole should not be added, or a special TIPL one. The doc covers everything.

**Comment by Tolriq** [comment]
@kgarner7 your wording is already very clear. Can you please update the PR without the duplicated comments so we can merge this?

File: content/en/docs/Responses/structuredLyrics.md
Comment (Tolriq) [reply content/en/docs/Responses/structuredLyrics.md:35]: @kgarner7 "synced" still present here.

**Review by Tolriq — DISMISSED** [review]

**Review by dweymouth — DISMISSED** [review]

**Comment by dweymouth** [comment]
I think this is pretty solid and uncontroversial - we just need a @opensubsonic/servers approval and this one can be good to go

**Review by gravelld — COMMENTED** [review]
A couple of things I noticed, looks good though.

File: content/en/docs/Responses/structuredLyrics.md
Comment (Tolriq) [reply content/en/docs/Responses/structuredLyrics.md:52]: You already have the song ID since you need it to request ;) Those values can be present in the lyrics themselves so are exposed

Comment (gravelld) [reply content/en/docs/Responses/structuredLyrics.md:52]: I don't really see the need for the displayX values at all then - I don't understand what you mean by:
> Those values can be present in the lyrics themselves

... but I'm an API minimalist so...

Comment (Tolriq) [reply content/en/docs/Responses/structuredLyrics.md:52]: As you can see they are optional ... And as said a lyric file can have headers that contains information, this is the information that is exposed here. If the information is available then it can be exposed.

File: content/en/docs/Responses/line.md
Comment (kgarner7) [reply content/en/docs/Responses/line.md:29]: Yes; feel free to suggest alternative wording if it would be clearer!

File: content/en/docs/Responses/structuredLyrics.md
Comment (kgarner7) [reply content/en/docs/Responses/structuredLyrics.md:52]: This was probably noted, but per [Wikioedia](https://en.wikipedia.org/wiki/LRC_(file_format) you can have fields like
```lrc
[ti:Lyrics (song) title]
[ar:Lyrics artist]
```
which could be the artist/song title, localized. Since those are not the same as the artist ID (and are only for display purposes), it would be nice to expose them (if they exist)

File: content/en/docs/Responses/line.md
Comment (gravelld) [reply content/en/docs/Responses/line.md:29]: What's our process here, can I just edit the file and place a commit?

Comment (kgarner7) [reply content/en/docs/Responses/line.md:29]: I think it probably makes sense to discuss here first (as any change will require everyone to approve again, so reducing excess commits is nice)

Comment (dweymouth) [reply content/en/docs/Responses/line.md:29]: The phrasing from @gravelld works for me

**Review by Tolriq — DISMISSED** [review]

**Review by dweymouth — APPROVED** [review]

**Review by gravelld — DISMISSED** [review]

**Comment by dweymouth** [comment]
Maybe we should give @deluan a chance to weigh in on this before merging since Navidrome is currently the OpenSubsonic server with the richest lyrics support

**Comment by deluan** [comment]
I don't see a reason why we need to parse and transform the lrc file into a structured json/xml, I think it adds complexities for both servers and clients, but if you guys see value in it, I'm good ok with that.

**Comment by dweymouth** [comment]
I think there is a lot of value in that - first and foremost it makes it clear to the client whether lyrics are synced or unsynced

**Comment by dweymouth** [comment]
I did just think of a complication though - if the language of lyrics are unknown what should be the server behavior @kgarner7 ? I would think unknown language lyrics should be returned if the server can't return lyrics known to be in the requested language, and leave the lang property as empty string on the returned lyrics object

**Comment by kgarner7** [comment]
> I did just think of a complication though - if the language of lyrics are unknown what should be the server behavior @kgarner7 ? I would think unknown language lyrics should be returned if the server can't return lyrics known to be in the requested language, and leave the lang property as empty string on the returned lyrics object

My understanding from the ID3v2.4 spec is that the language is a **required** part of the tag itself (they all have `Language  $xx xx xx`). This _should_ follow the ISO-639-2 spec, but I have also seen (and I believe most of my lyrics are) `XXX`. Not sure what to do about non-ISO language codes (I was expecting the query to just be by language code itself, which should be unambiguous).

**Comment by dweymouth** [comment]
Is lang required in .lrc files as well? I have a few albums that have lrc files, I guess I should take a look, though I'm away from having access to my server for awhile

**Comment by kgarner7** [comment]
Ah, no; I was just thinking of tags. While you _can_ add tags to lrc file, not required (and if a server reads a txt file, those are definitely not tagged). Not sure about how to handle that. XXX/Unknown?

**Comment by deluan** [comment]
> I think there is a lot of value in that - first and foremost it makes it clear to the client whether lyrics are synced or unsynced

I don't think this is enough reason to not use the LRC format, which is already structured and probably has libraries to support it in various programming languages. This would simplify clients that want to retrieve the lyrics from some online source directly (or from the local file system): They wouldn't need to implement two parsers.

For disambiguating between sync and unsynced, a field in the response payload would suffice, no?

Again, I'm late to the party and just giving my two cents. I'm ok with whatever was discussed and decided (specially because @kgarner7 will do the implementation ;) )

**Comment by dweymouth** [comment]
I would propose either "xxx" or empty string for unknown language. And also that the server should always include an unknown language result if it doesn't have a result matching the requested language(s). Then the client can decide if it wants to display it anyway or try its own sources if it has implemented external fetching

**Comment by kgarner7** [comment]
> > I think there is a lot of value in that - first and foremost it makes it clear to the client whether lyrics are synced or unsynced
> 
> I don't think this is enough reason to not use the LRC format, which is already structured and probably has libraries to support it in various programming languages. This would simplify clients that want to retrieve the lyrics from some online source directly (or from the local file system): They wouldn't need to implement two parsers.
> 
> For disambiguating between sync and unsynced, a field in the response payload would suffice, no?
> 
> Again, I'm late to the party and just giving my two cents. I'm ok with whatever was discussed and decided (specially because @kgarner7 will do the implementation ;) )

Some reasons why I proposed using the Jellyfin format:
- timestamp is unambiguous (I've seen issues where the "format" is `[mm:ss.xx]` but some sources return `[mm:ss.xxx]`; I guess we could be very strict here)
- makes newlines a bit easier (I guess the correct way in XML would be `&#xA;` for both synced/unsynced)
- I guess if someone supports both clients makes parsing more consistent

I don't really mind between Jellyfin format and text/LRC (with proper newlines for XML...). But if we do LRC, we must be strict on whether `[mm:ss.xxx]` or other formats are supported for the timestamp.

**Comment by deluan** [comment]
> Some reasons why I proposed using the Jellyfin format:

Ah, I didn't know this was an already used format. Then that makes total sense! Sorry about the noise :)

**Comment by kgarner7** [comment]
Yeah, the line format is inspired by https://github.com/jellyfin/jellyfin/pull/8381 

For lyrics, I think that if tags require 3-character lyrics, I would prefer storing lyrics with `XXX`. I guess the question then is, are we expecting the lyrics to actually abide by ISO standard? As far as I can tell, this is the last thing

**Comment by dweymouth** [comment]
We need to be agnostic to where the lyrics come from (tags, lrc files, Genius, etc) and that means we need to clarify and support the case of lyrics with unknown language

**Comment by Tolriq** [comment]
The common value is xxx when unknown for most cases.

And @deluan Emby and Plex also return structured data, this is actually a lot easier to handle than manual LRC parsing and all the edge cases + this allow support for other sources for the server that are not LRC files and avoid the need for the server to convert to a limited format.

**Comment by dweymouth** [comment]
@kgarner7 I think we just need to add then that unknown language lyrics are tagged "xxx" and *should* be returned if no lyrics for one of the requested languages are available.

**Comment by kgarner7** [comment]
> @kgarner7 I think we just need to add then that unknown language lyrics are tagged "xxx" and _should_ be returned if no lyrics for one of the requested languages are available.

Makes sense! Are there any other concerns by anyone else? I'd like to make this the last time we have to commit/get approval if possible. @dweymouth @Tolriq @deluan @gravelld (apologies for the pings)

**Comment by Tolriq** [comment]
You are very very far from what I had to do on the big PR :p

To avoid future remarks or discussion, maybe add a note somewhere that this is related to all lyrics kind (tags, external lrc files, external sources, ...)

**Comment by kgarner7** [comment]
Yes, but I am _trying_ to do things right here (probably should've pinged servers in the original discussion I guess...).

Agree on the second point, and I think an important clarification should be that servers must/should (not sure how strong to be here) be able to support multiple lyrics of the type (synced/unsynced) and language. For file tags this will never be an issue, but if servers support tags and external files/sources then you could have duplicates.

Alternatively, if we want to require only one, per language, can just say something like
> Servers must only have one lyric per type (synced/unsynced) and language code (including XXX/Unknown) combination. Selecting between duplicates is left as a server implementation detail.

**Comment by Tolriq** [comment]
No I don't think there should be any limit.
The user can put a few LRC files and test to find the proper one, faster if he sees the list than changing files everytime.
Something I do quite often with subtitles.

**Comment by kgarner7** [comment]
As it turns out, there is a specific code `und` (per https://www.loc.gov/standards/iso639-2/php/code_list.php) for unknown language, so that is the appropriate one.

**Review by dweymouth — APPROVED** [review]

**Comment by Tolriq** [comment]
If you change from xxx you must add a note that xxx should be replaced with und by the server. xxx is the defacto standard and what ffmpeg use when transcoding for example. So it will be present in many places.

**Comment by kgarner7** [comment]
.-. I guess, an alternative would be to treat both `und` (actual standard) and `xxx` as the special "undefined"/"unspecified", so if the server does not have a match but has one of these two it will return them, and then default to setting `xxx`. Or just converting `und` -> `xxx`. 

If people have recommendations/solutions, please share them (maybe make a discussion so it threads easier?). I would like a sane solution that isn't too onerous on the server, but _can_ (in theory) enable clients to use iso codes -> language where possible.

**Comment by Tolriq** [comment]
Honestly at this point since this is for a single song anyway and people having multiple lyrics with different languages are being the exception I would say that we drop the filter, return all and the clients deals with it.
That's the simplest for the servers and not really complicated for the clients to handle.

**Review by Tolriq — APPROVED** [review]

**Review by dweymouth — APPROVED** [review]

**Review by deluan — APPROVED** [review]

**Comment by Tolriq** [comment]
@epoupon @sentriz would be nice to have another server validation then I merge.

**Review by sentriz — APPROVED** [review]

**Comment by epoupon** [comment]
Hello!
I have a question on this.

If we have a LRC file containing:
```
[00:01.30]First line
Second line
[00:05.30] Third line
```

Are we supposed to have:
```
"line": [
            {
              "start": 1300,
              "value": "First line"
            },
            {
              "value": "Second line"
            },
            {
              "start": 5300,
              "value": "Third line"
            }
          ]
```

or:
```
"line": [
            {
              "start": 1300,
              "value": "First line\nSecond line"
            },
            {
              "start": 5300,
              "value": "Third line"
            }
          ]
```
According to the docs, I don't know. According to discussions here I would say likely option 2

**Comment by Tolriq** [comment]
Have you ever seen such file?

Most of the parsers just ignore the lines that do not have a valid time at the start to filter the headers and other comments sometimes present.

**Comment by paulijar** [comment]
I don't know what is the correct answer, but my server does what @Tolriq said and ignores the lines with no timestamp tag.

**Comment by kgarner7** [comment]
At this point, I don't remember the history but Navidrome does the second option (supporting multiline lyrics _if_ there's a timestamp before).

**Pull request merged**
